cpp/src/arrow/array/dict_internal.cc

```diff
@@ -44,119 +44,114 @@ using internal::CopyBitmap;
 // ----------------------------------------------------------------------
 // DictionaryType unification
 
-struct UnifyDictionaryValues {
-  MemoryPool* pool_;
-  std::shared_ptr<DataType> value_type_;
-  const std::vector<const DictionaryType*>& types_;
-  const std::vector<const Array*>& dictionaries_;
-  std::shared_ptr<Array>* out_values_;
-  std::vector<std::vector<int32_t>>* out_transpose_maps_;
-
-  template <typename T>
-  enable_if_no_memoize<T, Status> Visit(const T&) {
-    // Default implementation for non-dictionary-supported datatypes
-    return Status::NotImplemented("Unification of ", value_type_,
-                                  " dictionaries is not implemented");
+template <typename T>
+class DictionaryUnifierImpl : public DictionaryUnifier {
+ public:
+  using ArrayType = typename TypeTraits<T>::ArrayType;
+  using DictTraits = typename internal::DictionaryTraits<T>;
+  using MemoTableType = typename DictTraits::MemoTableType;
+
+  DictionaryUnifierImpl(MemoryPool* pool, std::shared_ptr<DataType> value_type)
+      : pool_(pool), value_type_(value_type), memo_table_(pool) {}
+
+  Status Unify(const Array& dictionary) override {
```

Comment:
Could simply pass `out = nullptr` to the other `Unify` method instead of duplicating the memo table insertion code.